clang/lib/Serialization/ASTReader.cpp

```diff
@@ -291,7 +291,30 @@ ASTReaderListener::~ASTReaderListener() = default;
 static bool checkLanguageOptions(const LangOptions &LangOpts,
                                  const LangOptions &ExistingLangOpts,
                                  DiagnosticsEngine *Diags,
-                                 bool AllowCompatibleDifferences = true) {
+                                 bool AllowCompatibleDifferences = true,
+                                 bool IsDependencyScanning = false) {
+  // For explicit modules workflow \c NeededByPCHOrCompilationUsesPCH is not a
+  // meaningful difference for the explicit modules and we canonicalize it away.
+  // Treat it as "benign" for this ASTReader check.
+  // NOTE: This can go away if we get rid of NeededByPCHOrCompilationUsesPCH.
+  bool IsExplicitModulesWorkflow =
+      IsDependencyScanning || !ExistingLangOpts.ImplicitModules;
+  bool ExistingNeededByPCHOrCompilationUsesPCH =
+      ExistingLangOpts.NeededByPCHOrCompilationUsesPCH;
+  if (IsExplicitModulesWorkflow) {
+    // Match NeededByPCHOrCompilationUsesPCH to make it "benign".
```

Comment:
I don't particularly like this hack but I wanted to minimize the differences from upstream when dealing with `NeededByPCHOrCompilationUsesPCH`.